BrainPortal/app/models/s3_flat_data_provider.rb

```diff
@@ -0,0 +1,299 @@
+    by_user.is_member_of_group(self.meta[:browse_gid].to_i)
+  end
+
+  def allow_file_owner_change? #:nodoc:
+    true
+  end
+
+  def content_storage_shared_between_users? #:nodoc:
+    true
+  end
+
+  def provider_full_path(userfile)
+    init_connection
+    @s3_connection.clean_starting_folder_path(userfile.name)
+  end
+
+  def impl_provider_collection_index(userfile, directory = :all, allowed_types = :regular) #:nodoc:
+    list = []
+    types = allowed_types.is_a?(Array) ? allowed_types.dup : [allowed_types]
+    types.map!(&:to_sym)
+
+    init_connection
+    entries = []
+    if userfile.is_a? FileCollection
+      if directory == :all
+        entriesTmp = s3_connection.list_objects_long(userfile.name)
+        entriesTmp.each do |e|
+          x = @s3_connection.get_object_stats(e[:key])
+          x[:name] = e[:key]
+          entries << x
+        end
+      else
+        if directory == :top or directory == "."# Think this indicates the whole bucket
+          entriesTmp = s3_connection.list_objects_short(userfile.name.to_s + "/")
+          directory_to_pass = ""
+        else
+          full_dir_name = File.join(userfile.name.to_s,directory).to_s + "/"
+          entriesTmp = s3_connection.list_objects_short(full_dir_name)
+          directory_to_pass = directory
+        end
+        fullPath = entriesTmp[:path]
+        entriesTmp[:files].each do |e|
+          fname = fullPath.to_s + e[:name]
+          entry = @s3_connection.get_object_stats(fname)
+          entry[:name] = directory_to_pass != "" ? File.join(directory_to_pass,e[:name]).to_s : e[:name]
+          entries << entry
+        end
+        entriesTmp[:folders].each do |d|
+          fname = fullPath.to_s + d[:name]
+          date_and_size = @s3_connection.get_mod_date_and_size_for_folder(fname)
+          entry = {:name => directory_to_pass != "" ? File.join(directory_to_pass,d[:name]).to_s : d[:name],
+                   :last_modified => date_and_size[:last_modified],
+                   :content_length => date_and_size[:content_length],
+                   :content_type => "application/x-directory"}
+          entries << entry
+        end
+      end
+    else
+      entry = @s3_connection.get_object_stats(provider_full_path(userfile))
+      entry[:name] = userfile.name
+      entries << entry
+    end
+
+    entries.each do |entry|
+      type = @s3_connection.translate_content_type_to_ftype(entry[:content_type])
+      next unless types.include?(type)
+      next if is_excluded?(entry[:name]) # in DataProvider
+      #next if entry[:name].to_s.begins_with("cbrain_hidden_")
+      fileinfo = FileInfo.new
+      fileinfo.name          = entry[:name]
+      fileinfo.symbolic_type = type
+      fileinfo.size          = entry[:content_length]
+      fileinfo.mtime         = entry[:last_modified]
+      fileinfo.owner         = "s3"
+      fileinfo.group         = "s3"
+      list << fileinfo
+   end
+   list.sort! { |a,b| a.name <=> b.name }
```

Comment:
Bugs:
* When asking for files, all returned entries are prefixed with the bucket name
* When asking for directories, always got an empty array